File: groovelab_backup_20260325/macos/Runner/AudioEngine/MetronomeEngine.swift
import AVFoundation

/// High-precision metronome engine for macOS using AVAudioEngine.
/// Port of the iOS MetronomeEngine — no AVAudioSession or haptics on macOS.
final class MetronomeEngine {
    // MARK: - Audio Graph
    private var audioEngine: AVAudioEngine?
    private var mixerNode: AVAudioMixerNode?
    private var buffers: [String: AVAudioPCMBuffer] = [:]

    // MARK: - Timing State (all atomic via queue)
    private let audioQueue = DispatchQueue(label: "com.groovelab.metronome", qos: .userInteractive)
    private var timer: DispatchSourceTimer?
    private var isRunning = false

    // Metronome parameters
    private var bpm: Int = 120
    private var beatsPerBar: Int = 4
    private var beatUnit: Int = 4
    private var subdivision: Int = 1
    private var swingPercent: Int = 0
    private var clickSound: String = "Wood"
    private var accentPattern: [Double] = [1.0, 0.7, 0.7, 0.7]
    private var humanFeel: Int = 0
    private var polyrhythmEnabled: Bool = false
    private var polyrhythmValue: Int = 3

    // Drum mode
    private var isDrumMode = false
    private var drumPattern: [String: [Int]] = [:]
    private var drumVolumes: [String: Double] = ["kick": 1.0, "snare": 1.0, "hihat": 1.0, "ride": 1.0]

    // Scheduling state
    private var nextBeatTime: AVAudioTime?
    private var currentSubBeat: Int = 0
    private var currentMeasure: Int = 0
    private var sampleRate: Double = 44100.0

    // Lookahead: schedule 30ms ahead, check every 10ms
    private let lookAheadMs: Double = 30.0
    private let timerIntervalMs: Double = 10.0

    // Callback to send events to Flutter
    private let onEvent: ([String: Any]) -> Void

    // Pool of player nodes for polyphonic playback
    private let playerPoolSize = 8
    private var playerPool: [AVAudioPlayerNode] = []
    private var nextPlayerIndex = 0

    init(onEvent: @escaping ([String: Any]) -> Void) {
        self.onEvent = onEvent
    }

    // MARK: - Initialization

    func initialize() {
        setupAudioEngine()
    }

    private func setupAudioEngine() {
        let engine = AVAudioEngine()
        let mixer = AVAudioMixerNode()
        engine.attach(mixer)

        // On macOS, get sample rate from the output node
        sampleRate = engine.outputNode.outputFormat(forBus: 0).sampleRate
        if sampleRate == 0 { sampleRate = 44100.0 }

        let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!

        for _ in 0..<playerPoolSize {
            let player = AVAudioPlayerNode()
            engine.attach(player)
            engine.connect(player, to: mixer, format: format)
            playerPool.append(player)
        }

        engine.connect(mixer, to: engine.mainMixerNode, format: format)

        do {
            try engine.start()
        } catch {
            print("AVAudioEngine start error: \(error)")
        }

        self.audioEngine = engine
        self.mixerNode = mixer
    }

    // MARK: - Sound Loading

    func loadSound(key: String, wavData: Data) {
        guard let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1) else { return }
        guard wavData.count > 44 else { return }
        let pcmData = wavData.subdata(in: 44..<wavData.count)
        let sampleCount = pcmData.count / 2

        guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: AVAudioFrameCount(sampleCount)) else { return }
        buffer.frameLength = AVAudioFrameCount(sampleCount)

        let floatData = buffer.floatChannelData![0]
        pcmData.withUnsafeBytes { (rawPtr: UnsafeRawBufferPointer) in
            let int16Ptr = rawPtr.bindMemory(to: Int16.self)
            for i in 0..<sampleCount {
                floatData[i] = Float(int16Ptr[i]) / 32768.0
            }
        }

        buffers[key] = buffer
    }

    // MARK: - Metronome Start/Stop

    func start(
        bpm: Int, beatsPerBar: Int, beatUnit: Int, subdivision: Int,
        swingPercent: Int, clickSound: String, accentPattern: [Double],
        hapticEnabled: Bool
    ) {
        audioQueue.async { [self] in
            stop_internal()
            self.bpm = bpm
            self.beatsPerBar = beatsPerBar
            self.beatUnit = beatUnit
            self.subdivision = subdivision
            self.swingPercent = swingPercent
            self.clickSound = clickSound
            self.accentPattern = accentPattern
            self.isDrumMode = false
            self.currentSubBeat = 0
            self.currentMeasure = 0

            guard let engine = audioEngine, engine.isRunning else { return }
            for player in playerPool { if !player.isPlaying { player.play() } }

            let hostNow = mach_absolute_time()
            var timebaseInfo = mach_timebase_info_data_t()
            mach_timebase_info(&timebaseInfo)
            let nsPerTick = Double(timebaseInfo.numer) / Double(timebaseInfo.denom)
            let offsetNs = UInt64(lookAheadMs * 1_000_000)
            nextBeatTime = AVAudioTime(hostTime: hostNow + UInt64(Double(offsetNs) / nsPerTick))

            isRunning = true
            startTimer()
        }
    }

    func startDrumPattern(bpm: Int, pattern: [String: [Int]], swingPercent: Int) {
        audioQueue.async { [self] in
            stop_internal()
            self.bpm = bpm
            self.beatsPerBar = 4
            self.beatUnit = 4
            self.subdivision = 4
            self.swingPercent = swingPercent
            self.drumPattern = pattern
            self.isDrumMode = true
            self.currentSubBeat = 0
            self.currentMeasure = 0

            guard let engine = audioEngine, engine.isRunning else { return }
            for player in playerPool { if !player.isPlaying { player.play() } }

            let hostNow = mach_absolute_time()
            var timebaseInfo = mach_timebase_info_data_t()
            mach_timebase_info(&timebaseInfo)
            let nsPerTick = Double(timebaseInfo.numer) / Double(timebaseInfo.denom)
            let offsetNs = UInt64(lookAheadMs * 1_000_000)
            nextBeatTime = AVAudioTime(hostTime: hostNow + UInt64(Double(offsetNs) / nsPerTick))

            isRunning = true
            startTimer()
        }
    }

    func stop() {
        audioQueue.async { [self] in stop_internal() }
    }

    private func stop_internal() {
        isRunning = false
        timer?.cancel()
        timer = nil
        for player in playerPool { player.stop() }
    }

    // MARK: - Timer / Scheduler

    private func startTimer() {
        timer?.cancel()
        let t = DispatchSource.makeTimerSource(flags: .strict, queue: audioQueue)
        let intervalNs = UInt64(timerIntervalMs * 1_000_000)
        t.schedule(deadline: .now(), repeating: .nanoseconds(Int(intervalNs)), leeway: .nanoseconds(0))
        t.setEventHandler { [weak self] in self?.schedulerTick() }
        t.resume()
        timer = t
    }

    private func schedulerTick() {
        guard isRunning, let engine = audioEngine, engine.isRunning else { return }

        var timebaseInfo = mach_timebase_info_data_t()
        mach_timebase_info(&timebaseInfo)
        let nsPerTick = Double(timebaseInfo.numer) / Double(timebaseInfo.denom)
        let now = mach_absolute_time()
        let nowNs = Double(now) * nsPerTick
        let lookAheadNs = lookAheadMs * 1_000_000

        while isRunning {
            guard let beatTime = nextBeatTime else { break }
            let beatNs = Double(beatTime.hostTime) * nsPerTick
            if beatNs > nowNs + lookAheadNs { break }

            if isDrumMode {
                scheduleDrumStep(at: beatTime)
            } else {
                scheduleMetronomeClick(at: beatTime)
            }
            advanceToNextBeat(nsPerTick: nsPerTick)
        }
    }

    private func scheduleMetronomeClick(at time: AVAudioTime) {
        let totalSubBeats = beatsPerBar * subdivision
        let subBeatInBar = currentSubBeat % totalSubBeats
        let isMainBeat = subBeatInBar % subdivision == 0
        let mainBeatIndex = subBeatInBar / subdivision

        if isMainBeat {
            let vol = mainBeatIndex < accentPattern.count ? accentPattern[mainBeatIndex] : 0.7
            if vol > 0 {
                let isAccent = vol >= 0.9
                let soundKey = getSoundKey(clickSound: clickSound, isAccent: isAccent)
                scheduleBuffer(key: soundKey, at: time)
            }

            var info = mach_timebase_info_data_t()
            mach_timebase_info(&info)
            let tsUs = Int(Double(time.hostTime) * Double(info.numer) / Double(info.denom) / 1000)
            onEvent([
                "type": "beat",
                "beatIndex": mainBeatIndex,
                "measureIndex": currentMeasure,
                "isAccent": mainBeatIndex == 0,
                "timestampUs": tsUs,
            ])
        } else if subdivision > 1 {
            scheduleBuffer(key: "click_sub", at: time)
        }

        if polyrhythmEnabled && !isDrumMode && isMainBeat {
            let totalBeats = beatsPerBar
            let polyN = polyrhythmValue
            let lcmVal = lcm(totalBeats, polyN)
            let polyStep = lcmVal / polyN
            let mainStep = lcmVal / totalBeats
            let currentPos = mainBeatIndex * mainStep
            if currentPos % polyStep == 0 {
                scheduleBuffer(key: "click_ghost", at: time, volume: 0.4)
            }
        }
    }

    private func scheduleDrumStep(at time: AVAudioTime) {
        let step = currentSubBeat % 16
        for track in ["kick", "snare", "hihat", "ride"] {
            guard let pattern = drumPattern[track], step < pattern.count, pattern[step] == 1 else { continue }
            let vol = drumVolumes[track] ?? 1.0
            if vol > 0.01 { scheduleBuffer(key: track, at: time, volume: Float(vol)) }
        }

        var info = mach_timebase_info_data_t()
        mach_timebase_info(&info)
        let tsUs = Int(Double(time.hostTime) * Double(info.numer) / Double(info.denom) / 1000)
        onEvent(["type": "drumStep", "step": step, "measureIndex": currentMeasure, "timestampUs": tsUs])
    }

    private func scheduleBuffer(key: String, at time: AVAudioTime, volume: Float = 1.0) {
        guard let buffer = buffers[key] else { return }
        let player = playerPool[nextPlayerIndex % playerPoolSize]
        nextPlayerIndex += 1
        if !player.isPlaying { player.play() }
        player.volume = volume
        player.scheduleBuffer(buffer, at: time, options: [], completionHandler: nil)
    }

    private func advanceToNextBeat(nsPerTick: Double) {
        let totalSubBeats = isDrumMode ? 16 : (beatsPerBar * subdivision)
        currentSubBeat += 1
        if currentSubBeat % totalSubBeats == 0 { currentMeasure += 1 }

        let baseIntervalSec: Double
        if isDrumMode {
            baseIntervalSec = 60.0 / Double(bpm) / 4.0
        } else {
            baseIntervalSec = 60.0 / Double(bpm) / Double(subdivision)
        }

        var intervalNs = baseIntervalSec * 1_000_000_000

        if swingPercent > 0 && subdivision >= 2 {
            let swingRatio = 0.5 + Double(swingPercent) / 200.0
            let pairInterval = baseIntervalSec * 2.0
            let isEvenSubBeat = (currentSubBeat - 1) % 2 == 0
            intervalNs = (isEvenSubBeat ? pairInterval * swingRatio : pairInterval * (1.0 - swingRatio)) * 1_000_000_000
        }

        if humanFeel > 0 && !isDrumMode {
            let maxJitterFraction = Double(humanFeel) / 100.0 * 0.08
            let jitter = Double.random(in: -1.0...1.0) * maxJitterFraction * intervalNs
            intervalNs += jitter
            intervalNs = max(intervalNs * 0.85, intervalNs)
        }

        guard let prev = nextBeatTime else { return }
        let advanceTicks = UInt64(intervalNs / nsPerTick)
        nextBeatTime = AVAudioTime(hostTime: prev.hostTime + advanceTicks)
    }

    private func getSoundKey(clickSound: String, isAccent: Bool) -> String {
        switch clickSound {
        case "Wood": return isAccent ? "click_accent" : "click_normal"
        case "Digital": return isAccent ? "digital_accent" : "digital_normal"
        case "Hi-Hat": return "hihat_click"
        case "Clave": return isAccent ? "clave_accent" : "clave_normal"
        case "Cowbell": return isAccent ? "cowbell_accent" : "cowbell_normal"
        case "Beep": return isAccent ? "beep_accent" : "beep_normal"
        default: return isAccent ? "click_accent" : "click_normal"
        }
    }

    // MARK: - Live Updates
    func updateBpm(_ newBpm: Int) { audioQueue.async { self.bpm = newBpm } }
    func updateTimeSignature(beatsPerBar: Int, beatUnit: Int) {
        audioQueue.async {
            self.beatsPerBar = beatsPerBar
            self.beatUnit = beatUnit
            if self.accentPattern.count != beatsPerBar {
                var newAccents = Array(repeating: 0.7, count: beatsPerBar)
                newAccents[0] = 1.0
                self.accentPattern = newAccents
            }
        }
    }
    func updateSubdivision(_ sub: Int) { audioQueue.async { self.subdivision = sub } }
    func updateSwing(_ pct: Int) { audioQueue.async { self.swingPercent = pct } }
    func updateClickSound(_ sound: String) { audioQueue.async { self.clickSound = sound } }
    func updateAccentPattern(_ pattern: [Double]) { audioQueue.async { self.accentPattern = pattern } }
    func setHapticMode(_ enabled: Bool) { /* No haptics on macOS */ }
    func updateHumanFeel(_ percent: Int) { audioQueue.async { self.humanFeel = min(50, max(0, percent)) } }
    func updatePolyrhythm(enabled: Bool, value: Int) {
        audioQueue.async { self.polyrhythmEnabled = enabled; self.polyrhythmValue = max(2, min(7, value)) }
    }
    func updateDrumPattern(_ pattern: [String: [Int]]) { audioQueue.async { self.drumPattern = pattern } }
    func updateDrumVolumes(_ volumes: [String: Double]) { audioQueue.async { self.drumVolumes = volumes } }

    private func gcd(_ a: Int, _ b: Int) -> Int { b == 0 ? a : gcd(b, a % b) }
    private func lcm(_ a: Int, _ b: Int) -> Int { a / gcd(a, b) * b }

    func getOutputLatency() -> Double {
        // macOS: estimate from audio engine
        return (audioEngine?.outputNode.presentationLatency ?? 0) * 1000.0
    }
}

File: groovelab_backup_20260325/android/app/src/main/kotlin/com/groovelab/groovelab/audio/MetronomeEngine.kt
package com.groovelab.groovelab.audio

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.content.Context
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

/**
 * High-precision metronome engine for Android using AudioTrack.
 * Runs audio generation on a dedicated high-priority thread.
 * Uses System.nanoTime() for precise scheduling with drift compensation.
 */
class MetronomeEngine(private val onEvent: (Map<String, Any>) -> Unit) {

    private val sampleRate = 44100
    private val soundBuffers = mutableMapOf<String, ShortArray>()

    // Metronome state (volatile for thread safety)
    @Volatile private var bpm = 120
    @Volatile private var beatsPerBar = 4
    @Volatile private var beatUnit = 4
    @Volatile private var subdivision = 1
    @Volatile private var swingPercent = 0
    @Volatile private var clickSound = "Wood"
    @Volatile private var accentPattern = listOf(1.0, 0.7, 0.7, 0.7)
    @Volatile private var hapticEnabled = false
    @Volatile private var humanFeel = 0 // 0-50, timing jitter percentage
    @Volatile private var polyrhythmEnabled = false
    @Volatile private var polyrhythmValue = 3 // N in N:beatsPerBar

    // Drum mode
    @Volatile private var isDrumMode = false
    @Volatile private var drumPattern = mapOf<String, List<Int>>()
    @Volatile private var drumVolumes = mapOf("kick" to 1.0, "snare" to 1.0, "hihat" to 1.0, "ride" to 1.0)

    // Playback
    private var audioTrack: AudioTrack? = null
    private var playbackThread: Thread? = null
    @Volatile private var isRunning = false
    private var currentSubBeat = 0
    private var currentMeasure = 0

    fun initialize() {
        // Pre-allocate AudioTrack
        val minBuffer = AudioTrack.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )

        val attrs = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
            .build()

        val format = AudioFormat.Builder()
            .setSampleRate(sampleRate)
            .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
            .build()

        val builder = AudioTrack.Builder()
            .setAudioAttributes(attrs)
            .setAudioFormat(format)
            .setBufferSizeInBytes(minBuffer * 2)
            .setTransferMode(AudioTrack.MODE_STREAM)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder.setPerformanceMode(AudioTrack.PERFORMANCE_MODE_LOW_LATENCY)
        }

        audioTrack = builder.build()
    }

    fun loadSound(key: String, wavData: ByteArray) {
        // Parse WAV: skip 44-byte header, extract 16-bit PCM samples
        if (wavData.size <= 44) return
        val pcmData = wavData.copyOfRange(44, wavData.size)
        val sampleCount = pcmData.size / 2
        val samples = ShortArray(sampleCount)

        val buffer = ByteBuffer.wrap(pcmData).order(ByteOrder.LITTLE_ENDIAN)
        for (i in 0 until sampleCount) {
            samples[i] = buffer.short
        }
        soundBuffers[key] = samples
    }

    /** Play a single sound hit immediately (for manual drum pad taps). */
    fun playSingleHit(key: String, volume: Float = 1.0f) {
        val samples = soundBuffers[key] ?: return
        Thread {
            try {
                val track = audioTrack ?: return@Thread
                if (track.playState != AudioTrack.PLAYSTATE_PLAYING) {
                    track.play()
                }
                track.write(samples, 0, samples.size)
            } catch (e: Exception) {
                android.util.Log.e("MetronomeEngine", "playSingleHit error: $e")
            }
        }.start()
    }

    fun start(
        bpm: Int, beatsPerBar: Int, beatUnit: Int, subdivision: Int,
        swingPercent: Int, clickSound: String, accentPattern: List<Double>,
        hapticEnabled: Boolean
    ) {
        stop()
        this.bpm = bpm
        this.beatsPerBar = beatsPerBar
        this.beatUnit = beatUnit
        this.subdivision = subdivision
        this.swingPercent = swingPercent
        this.clickSound = clickSound
        this.accentPattern = accentPattern
        this.hapticEnabled = hapticEnabled
        this.isDrumMode = false
        this.currentSubBeat = 0
        this.currentMeasure = 0

        isRunning = true
        audioTrack?.play()
        startPlaybackThread()
    }

    fun startDrumPattern(bpm: Int, pattern: Map<String, List<Int>>, swingPercent: Int) {
        stop()
        this.bpm = bpm
        this.beatsPerBar = 4
        this.beatUnit = 4
        this.subdivision = 4
        this.swingPercent = swingPercent
        this.drumPattern = pattern
        this.isDrumMode = true
        this.currentSubBeat = 0
        this.currentMeasure = 0

        isRunning = true
        audioTrack?.play()
        startPlaybackThread()
    }

    fun stop() {
        isRunning = false
        playbackThread?.join(500)
        playbackThread = null
        audioTrack?.pause()
        audioTrack?.flush()
    }

    /**
     * Core playback thread: runs at high priority.
     * Instead of using Timer, this thread generates audio directly
     * and writes to AudioTrack, which handles precise timing.
     */
    private fun startPlaybackThread() {
        playbackThread = Thread({
            android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_URGENT_AUDIO)

            var nextBeatTimeNs = System.nanoTime()

            while (isRunning) {
                val currentBpm = bpm
                val currentSub = subdivision
                val totalSubBeats = if (isDrumMode) 16 else (beatsPerBar * currentSub)

                // Calculate this sub-beat's interval
                val baseIntervalNs = if (isDrumMode) {
                    (60_000_000_000L / currentBpm / 4) // 16th notes
                } else {
                    (60_000_000_000L / currentBpm / currentSub)
                }

                var intervalNs = baseIntervalNs

                // Apply swing
                if (swingPercent > 0 && currentSub >= 2) {
                    val swingRatio = 0.5 + swingPercent / 200.0
                    val pairIntervalNs = baseIntervalNs * 2
                    val isEven = (currentSubBeat) % 2 == 0
                    intervalNs = if (isEven) {
                        (pairIntervalNs * swingRatio).toLong()
                    } else {
                        (pairIntervalNs * (1.0 - swingRatio)).toLong()
                    }
                }

                // Apply human feel (timing jitter)
                if (humanFeel > 0 && !isDrumMode) {
                    val maxJitterFraction = humanFeel / 100.0 * 0.08
                    val jitter = (Math.random() * 2.0 - 1.0) * maxJitterFraction * intervalNs
                    intervalNs = max((intervalNs * 0.85).toLong(), intervalNs + jitter.toLong())
                }

                // Generate audio for this beat
                val soundData = generateBeatAudio()

                // Write audio to AudioTrack (this blocks until the buffer is consumed = natural timing)
                if (soundData != null && soundData.isNotEmpty()) {
                    audioTrack?.write(soundData, 0, soundData.size)
                }

                // Generate silence to fill remaining interval
                val soundDurationSamples = soundData?.size ?: 0
                val intervalSamples = (intervalNs.toDouble() / 1_000_000_000.0 * sampleRate).toInt()
                val silenceSamples = max(0, intervalSamples - soundDurationSamples)
                if (silenceSamples > 0) {
                    val silence = ShortArray(silenceSamples)
                    audioTrack?.write(silence, 0, silence.size)
                }

                // Send events
                if (isDrumMode) {
                    val step = currentSubBeat % 16
                    onEvent(mapOf(
                        "type" to "drumStep",
                        "step" to step,
                        "measureIndex" to currentMeasure,
                        "timestampUs" to (System.nanoTime() / 1000)
                    ))
                } else {
                    val subBeatInBar = currentSubBeat % totalSubBeats
                    val isMainBeat = subBeatInBar % currentSub == 0
                    if (isMainBeat) {
                        val mainBeatIndex = subBeatInBar / currentSub
                        onEvent(mapOf(
                            "type" to "beat",
                            "beatIndex" to mainBeatIndex,
                            "measureIndex" to currentMeasure,
                            "isAccent" to (mainBeatIndex == 0),
                            "timestampUs" to (System.nanoTime() / 1000)
                        ))
                    }
                }

                // Advance
                currentSubBeat++
                if (currentSubBeat % totalSubBeats == 0) {
                    currentMeasure++
                }
            }
        }, "GrooveLab-Audio").apply {
            priority = Thread.MAX_PRIORITY
        }
        playbackThread?.start()
    }

    private fun generateBeatAudio(): ShortArray? {
        if (isDrumMode) {
            return generateDrumStepAudio()
        }

        val totalSubBeats = beatsPerBar * subdivision
        val subBeatInBar = currentSubBeat % totalSubBeats
        val isMainBeat = subBeatInBar % subdivision == 0
        val mainBeatIndex = subBeatInBar / subdivision

        var mainSound: ShortArray? = null

        if (isMainBeat) {
            val vol = if (mainBeatIndex < accentPattern.size) accentPattern[mainBeatIndex] else 0.7
            if (vol > 0) {
                val isAccent = vol >= 0.9
                val key = getSoundKey(clickSound, isAccent)
                mainSound = soundBuffers[key]
            }

            // Polyrhythm: mix ghost click if this beat aligns with the poly grid
            if (polyrhythmEnabled) {
                val lcmVal = lcm(beatsPerBar, polyrhythmValue)
                val polyStep = lcmVal / polyrhythmValue
                val mainStep = lcmVal / beatsPerBar
                val currentPos = mainBeatIndex * mainStep
                if (currentPos % polyStep == 0) {
                    val ghost = soundBuffers["click_ghost"]
                    if (ghost != null && mainSound != null) {
                        mainSound = mixSounds(mainSound!!, ghost, 0.4)
                    } else if (ghost != null) {
                        mainSound = applyVolume(ghost, 0.4)
                    }
                }
            }
        } else if (subdivision > 1) {
            mainSound = soundBuffers["click_sub"]
        }

        return mainSound
    }

    private fun mixSounds(a: ShortArray, b: ShortArray, bVol: Double): ShortArray {
        val maxLen = max(a.size, b.size)
        val result = ShortArray(maxLen)
        for (i in 0 until maxLen) {
            val sa = if (i < a.size) a[i].toInt() else 0
            val sb = if (i < b.size) (b[i].toDouble() * bVol).toInt() else 0
            result[i] = max(-32768, min(32767, sa + sb)).toShort()
        }
        return result
    }

    private fun applyVolume(data: ShortArray, vol: Double): ShortArray {
        return ShortArray(data.size) { max(-32768, min(32767, (data[it].toDouble() * vol).toInt())).toShort() }
    }

    private fun generateDrumStepAudio(): ShortArray? {
        val step = currentSubBeat % 16
        val trackNames = listOf("kick", "snare", "hihat", "ride")
        val sounds = mutableListOf<Pair<ShortArray, Double>>() // sound + volume

        for (track in trackNames) {
            val pattern = drumPattern[track] ?: continue
            if (step < pattern.size && pattern[step] == 1) {
                val buf = soundBuffers[track] ?: continue
                val vol = drumVolumes[track] ?: 1.0
                if (vol > 0.01) sounds.add(Pair(buf, vol))
            }
        }

        if (sounds.isEmpty()) return null

        // Mix all simultaneous drum sounds with per-track volume
        val maxLen = sounds.maxOf { it.first.size }
        val mixed = ShortArray(maxLen)
        for ((sound, vol) in sounds) {
            for (i in sound.indices) {
                val scaled = (sound[i].toDouble() * vol).toInt()
                val sum = mixed[i].toInt() + scaled
                mixed[i] = max(-32768, min(32767, sum)).toShort()
            }
        }
        return mixed
    }

    private fun getSoundKey(clickSound: String, isAccent: Boolean): String {
        return when (clickSound) {
            "Wood" -> if (isAccent) "click_accent" else "click_normal"
            "Digital" -> if (isAccent) "digital_accent" else "digital_normal"
            "Hi-Hat" -> "hihat_click"
            "Clave" -> if (isAccent) "clave_accent" else "clave_normal"
            "Cowbell" -> if (isAccent) "cowbell_accent" else "cowbell_normal"
            "Beep" -> if (isAccent) "beep_accent" else "beep_normal"
            else -> if (isAccent) "click_accent" else "click_normal"
        }
    }

    // Live updates
    fun updateBpm(newBpm: Int) { bpm = newBpm }
    fun updateTimeSignature(beats: Int, unit: Int) {
        beatsPerBar = beats; beatUnit = unit
        if (accentPattern.size != beats) {
            accentPattern = List(beats) { if (it == 0) 1.0 else 0.7 }
        }
    }
    fun updateSubdivision(sub: Int) { subdivision = sub }
    fun updateSwing(pct: Int) { swingPercent = pct }
    fun updateClickSound(sound: String) { clickSound = sound }
    fun updateAccentPattern(pattern: List<Double>) { accentPattern = pattern }
    fun setHapticMode(enabled: Boolean) { hapticEnabled = enabled }
    fun updateHumanFeel(percent: Int) { humanFeel = percent.coerceIn(0, 50) }
    fun updatePolyrhythm(enabled: Boolean, value: Int) {
        polyrhythmEnabled = enabled
        polyrhythmValue = value.coerceIn(2, 7)
    }
    fun updateDrumPattern(pattern: Map<String, List<Int>>) { drumPattern = pattern }
    fun updateDrumVolumes(volumes: Map<String, Double>) { drumVolumes = volumes }

    private fun gcd(a: Int, b: Int): Int = if (b == 0) a else gcd(b, a % b)
    private fun lcm(a: Int, b: Int): Int = a / gcd(a, b) * b

    fun getOutputLatency(): Double {
        // AudioTrack latency is approximately buffer size / sample rate
        val minBuffer = AudioTrack.getMinBufferSize(sampleRate, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT)
        return (minBuffer.toDouble() / sampleRate / 2) * 1000.0 // approximate ms
    }
}
